index.js

```diff
@@ -1,15 +1,48 @@
 'use strict'
 
 const fp = require('fastify-plugin')
+const xtend = require('xtend')
+
 const Ajv = require('ajv')
+const ajv = new Ajv({ removeAdditional: true, useDefaults: true, coerceTypes: true })
+
+const optsSchema = {
+  type: 'object',
+  required: [ 'schema' ],
+  properties: {
+    schema: { type: 'object', additionalProperties: true },
+    confKey: { type: 'string', default: 'config' },
+    data: {
+      oneOf: [
+        { type: 'array', items: { type: 'object' }, minItems: 1 },
+        { type: 'object' }
+      ],
+      default: {}
+    },
+    env: { type: 'boolean', default: true }
+  }
+}
+const optsSchemaValidator = ajv.compile(optsSchema)
 
 function loadAndValidateEnvironment (fastify, opts, done) {
-  const ajv = new Ajv({ removeAdditional: true, useDefaults: true, coerceTypes: true })
+  const isOptionValid = optsSchemaValidator(opts)
+  if (!isOptionValid) {
+    return done(new Error(optsSchemaValidator.errors.map(e => e.message)))
+  }
+
   const schema = opts.schema
   schema.additionalProperties = false
-  const confKey = opts.confKey || 'config'
+  const confKey = opts.confKey
+
+  let data = opts.data
```

Comment:
Probably here we want to remove the references of `data` since we are modifying it.
This will do the job :)
```js
let data = Object.assign({}, opts.data)
```

Reply:
Uhm... `opts.data` can be an array too.

does `JSON.parse(JSON.stringiy(opts.data))` work?

Reply:
I've just figure out that `Object.assign` as well as `JSON.parse(JSON.stringify(opts.data))` will remove all the references to object like the following:

```js
function Test () {
  this.hello = 'world'
}

Test.prototype.log = function (log) {
  console.log(log)
}
```

So for the moment we could keep as it is, and if in the future this will cause some issue we'll figure out :)

Reply:
data can be a array:
```js
> Object.assign({}, ['pippo'])
{ '0': 'pippo' }
```

Anyway, if you occurs in this error, feel free to PR! (and thanks a lot 😄 )